x-pack/legacy/plugins/apm/public/components/app/ServiceMap/index.tsx

```diff
@@ -37,37 +54,159 @@ ${theme.euiColorLightShade}`,
+  const openToast = useRef<string | null>(null);
+
+  const [responses, setResponses] = useState<ServiceMapAPIResponse[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [percentageLoaded, setPercentageLoaded] = useState(0);
+  const [, _setUnusedState] = useState(false);
+
+  const elements = useMemo(() => getCytoscapeElements(responses, search), [
+    responses,
+    search
+  ]);
+
+  const forceUpdate = useCallback(() => _setUnusedState(value => !value), []);
+
+  const getNext = useCallback(
+    async (input: { reset?: boolean; after?: string | undefined }) => {
+      const { start, end, uiFilters: strippedUiFilters, ...query } = params;
+
+      if (input.reset) {
+        renderedElements.current = [];
+        setResponses([]);
+      }
 
-  const { data } = useFetcher(
-    callApmApi => {
       if (start && end) {
-        return callApmApi({
-          pathname: '/api/apm/service-map',
-          params: { query: { start, end } }
-        });
+        setIsLoading(true);
+        try {
+          const data = await callApmApi({
+            pathname: '/api/apm/service-map',
+            params: {
+              query: {
+                ...query,
+                start,
+                end,
+                uiFilters: JSON.stringify(strippedUiFilters),
+                after: input.after
+              }
+            }
+          });
+          setResponses(resp => resp.concat(data));
+          setIsLoading(false);
+
+          const shouldGetNext =
+            responses.length + 1 < MAX_REQUESTS && data.after;
+
+          if (shouldGetNext) {
+            setPercentageLoaded(value => value + 30); // increase loading bar 30%
+            await getNext({ after: data.after });
+          }
+        } catch (error) {
+          setIsLoading(false);
+          notifications.toasts.addError(error, {
+            title: i18n.translate('xpack.apm.errorServiceMapData', {
+              defaultMessage: `Error loading service connections`
+            })
+          });
+        }
       }
     },
-    [start, end]
+    [callApmApi, params, responses.length, notifications.toasts]
   );
 
-  const elements = Array.isArray(data) ? data : [];
-  const license = useLicense();
+  useEffect(() => {
+    const loadServiceMaps = async () => {
+      setPercentageLoaded(5);
+      await getNext({ reset: true });
+      setPercentageLoaded(100);
+    };
+
+    loadServiceMaps();
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [params]);
+
+  useEffect(() => {
+    if (renderedElements.current.length === 0) {
+      renderedElements.current = elements;
+      return;
+    }
+
+    const newElements = elements.filter(element => {
+      return !find(renderedElements.current, el => isEqual(el, element));
+    });
+
+    const updateMap = () => {
+      renderedElements.current = elements;
+      if (openToast.current) {
+        notifications.toasts.remove(openToast.current);
+      }
```

Comment:
Do we still need to keep track of the currently open toast? Would be nice if we can remove this and just let the toast handle its own state.